File: CaloriesApi/src/main/java/com/example/CaloriesApi/Services/CaloriesCalculator.java

package com.example.CaloriesApi.Services;

import java.util.PriorityQueue;

import com.example.CaloriesApi.Models.CaloriesResult;
 

public class CaloriesCalculator {
    public static CaloriesResult getTopCalories(String data, int k) {
        String[] lines = data.split("\\r?\\n");

            PriorityQueue<Integer> topCalories = new PriorityQueue<>();
            int currentElfCalories = 0;
            int maxCalories = Integer.MIN_VALUE;

            for (String line : lines) {
                line = line.trim();
                    if (!line.isEmpty()) {
                        try {
                            currentElfCalories += Integer.parseInt(line);
                        } catch (NumberFormatException e) {
                            return new CaloriesResult(0, 0, "Invalid input: '" + line + "' is not a valid number.");
                        }
                    } else {
                        maxCalories = Math.max(maxCalories, currentElfCalories);
                        addElfCalories(topCalories, currentElfCalories, k);
                        currentElfCalories = 0;
                    }
            }

            maxCalories = Math.max(maxCalories, currentElfCalories);
            addElfCalories(topCalories, currentElfCalories, k);

            int topKCaloriesSum = 0;
            for (int calories : topCalories) {
                topKCaloriesSum += calories;
            }

            return  new CaloriesResult(maxCalories, topKCaloriesSum, "Success");
    }

    private static void addElfCalories(PriorityQueue<Integer> pq, int calories, int k) {
        if (calories > 0) {
            pq.add(calories);
            if (pq.size() > k) {
                pq.poll();
            }
        }
    }
}

File: frontend/src/utils/CaloriesCalculator.ts
export interface CaloriesInput {
  calories: string;
  k: number;
}

export interface CaloriesOutput {
  maxCalories: number;
  topKCaloriesSum: number;
  message?: string;
}

export const getTopCalories = (input: CaloriesInput): CaloriesOutput => {
  const totalCalories: number[] = [];
  let elfCalories = 0;

  const lines = input.calories.split(/\r?\n/);

  try {
    for (let line of lines) {
      if (line.trim()) {
        const num = Number(line.trim());

        if (isNaN(num)) {
          throw new Error(
            `Invalid input: '${line.trim()}' is not a valid number.`
          );
        }

        elfCalories += num;
      } else {
        if (elfCalories > 0) {
          totalCalories.push(elfCalories);
        }
        elfCalories = 0;
      }
    }

    if (elfCalories > 0) {
      totalCalories.push(elfCalories);
    }

    totalCalories.sort((a, b) => b - a);

    const maxCalories = totalCalories[0] || 0;

    const topKCaloriesSum = totalCalories
      .slice(0, input.k)
      .reduce((a, b) => a + b, 0);

    return {
      maxCalories,
      topKCaloriesSum,
      message: "Success",
    };
  } catch (error) {
    return {
      maxCalories: 0,
      topKCaloriesSum: 0,
      message: (error as Error).message,
    };
  }
};
